src/high-level/matrix.lisp

```diff
@@ -310,24 +310,91 @@ Target cannot be the same as a or b.")
 
 ;;; Generic matrix methods
 
+(define-extensible-function (block-diag block-diag-lisp) (mat &rest blocks)
+  (:documentation "Construct a matrix from its diagonal blocks.")
+  (:method (mat &rest blocks)
+      (push mat blocks)
+    (let ((nrows 0)
+          (ncols 0)
+          (type (element-type mat)))
+      (dolist (mat blocks)
+        (incf nrows (nrows mat))
+        (incf ncols (ncols mat))
+        ;; TODO: type coercion? for now we can be strict
+        (unless (equalp type (element-type mat))
+          (error "Unable to construct block matrix from blocks with disagreeing types.")))
+      (let ((result (zeros (list nrows ncols) :type type)))
+        (loop :with i := 0
+              :with j := 0
+              :for mat :in blocks
+              :do (slice-to mat (list 0 0) (shape mat) result (list i j))
+                  (incf i (nrows mat))
+                  (incf j (ncols mat)))
+        result))))
+
+(define-extensible-function (block-matrix block-matrix-lisp) (blocks shape)
```

Comment:
Given the use of `apply`, I suppose the implicit assumption is that `blocks` won't be so large that it might blow the stack.

Reply:
(You might consider `hstack` and `vstack` to take a list instead.)

Reply:
Good idea.

Reply:
I still kept `block-diag` taking rest arguments, because I think it's nice for it to be "light to the touch", but since `hstack` and `vstack` are intended to be, relatively speaking, lower level constructors the list argument makes sense for them. Given the scaling of block-matrix (N arguments of size M each yields a (NM)^2 matrix) I'm not so worried about blowing the stack for `block-diag`.

Reply:
I'm just doing this across the board (e.g. with `block-diag` too), so that we won't have to remember whether to use rest args or a list argument.